### PR TITLE
MLP - Multiclass Classifier

### PR DESCRIPTION
**Pull request by MikeLippincott**
The end goal, in a separate PR, is to merge the helper functions in to one file for both Binary and Multi class classifiers. I want to keep them separate in the interim for ease of troubleshooting and easy edits. I am open to suggestions as always.

For ease of reviewing I can combine the files now instead of a later PR, as a lot is redundant if this is easier please let me know and I can add a commit.

### CODE REVIEW

File: 2.MLP_Multi_Classifier/MLP_utils/utils.py [hunk 1]
```diff
@@ -0,0 +1,938 @@
...
+    params.TRAIN_EPOCHS = int(config["DEFAULT"]["TRAIN_EPOCHS"])
+    params.MIN_LAYERS = int(config["DEFAULT"]["MIN_LAYERS"])
+    params.MAX_LAYERS = int(config["DEFAULT"]["MAX_LAYERS"])
+    params.LAYER_UNITS_MIN = int(config["DEFAULT"]["LAYER_UNITS_MIN"])
+    params.LAYER_UNITS_MAX = int(config["DEFAULT"]["LAYER_UNITS_MAX"])
+    params.DROPOUT_MIN = float(config["DEFAULT"]["DROPOUT_MIN"])
+    params.DROPOUT_MAX = float(config["DEFAULT"]["DROPOUT_MAX"])
+    params.DROP_OUT_STEP = float(config["DEFAULT"]["DROP_OUT_STEP"])
+    params.LEARNING_RATE_MIN = float(config["DEFAULT"]["LEARNING_RATE_MIN"])
+    params.LEARNING_RATE_MAX = float(config["DEFAULT"]["LEARNING_RATE_MAX"])
```
Comment (d33bs) [inline 2.MLP_Multi_Classifier/MLP_utils/utils.py:70]: Seeing how these params are casted to specific data types, would there be an opportunity to use dataclass type specifications? ([see this reference](https://docs.python.org/3/library/dataclasses.html#dataclasses.field))

File: 2.MLP_Multi_Classifier/MLP_utils/utils.py [hunk 2]
```diff
@@ -0,0 +1,938 @@
...
+from sklearn.metrics import (
+    accuracy_score,
+    auc,
+    classification_report,
+    confusion_matrix,
+    f1_score,
+    precision_score,
+    recall_score,
+    roc_auc_score,
+    roc_curve,
```
Comment (d33bs) [inline 2.MLP_Multi_Classifier/MLP_utils/utils.py:29]: Were all of these imports used in this file? If not, consider removing ones which were not used.

File: 2.MLP_Multi_Classifier/MLP_utils/utils.py [hunk 3]
```diff
@@ -0,0 +1,938 @@
...
+    n_classes = OUT_FEATURES  # the number of classes in your multi-class problem
+
+    # Binarize the labels for the multiclass ROC calculation
+    y_true_binarized = label_binarize(test_data, classes=np.arange(n_classes))
+    y_score_binarized = label_binarize(prediction_list, classes=np.arange(n_classes))
+
+    # Compute ROC curve and ROC area for each class
+    fpr = dict()
+    tpr = dict()
+    roc_auc = dict()
```
Comment (d33bs) [inline 2.MLP_Multi_Classifier/MLP_utils/utils.py:897]: Consider creating these dictionaries by using `{}` instead of `dict()`.

File: 2.MLP_Multi_Classifier/MLP_utils/utils.py [hunk 4]
```diff
@@ -0,0 +1,938 @@
...
+    sns.lineplot(x=x, y=y2, data=df)
+
+
+def test_optimized_model(
+    model: torch.nn.Sequential,
+    test_loader: torch.utils.data.DataLoader,
+    in_features: int,
+    out_features: int,
+    parameter_dict: dict,
+    params: object,
```
Comment (d33bs) [inline 2.MLP_Multi_Classifier/MLP_utils/utils.py:778]: Consider renaming this parameter for greater specificity and possibly referencing the the related dataclass for the type hint (instead of a generic object). This comment applies to several functions in this file (see areas where `params: object` is used).

File: 2.MLP_Multi_Classifier/MLP_utils/utils.py [hunk 5]
```diff
@@ -0,0 +1,938 @@
...
+        if trial.should_prune():
+            raise optuna.exceptions.TrialPruned()
+
+    # I want information returned but only 1 metric required for the optimize function called by study.optimize
+    # with out this conditional statement the optimization will fail
+    if return_info == True:
+        print(f"Validation Accuracy: {np.mean(valid_acc)}")
+        print(f"Validation Loss: {np.mean(valid_loss)}")
+        print(f"Training Accuracy: {np.mean(train_acc)}")
+        print(f"Training Loss: {np.mean(train_loss)}")
```
Comment (d33bs) [inline 2.MLP_Multi_Classifier/MLP_utils/utils.py:540]: Seeing that this if branch does not make a return and the others below do, consider revising this to keep consistent with returns for the function (either returning something or not for all branches / cases). Doing this would help improve readability and expectations within the function. Please use your best judgement on this comment; this is more of a recommendation than strict rule.

File: 2.MLP_Multi_Classifier/MLP_utils/utils.py [hunk 6]
```diff
@@ -0,0 +1,938 @@
...
+    elif metric == "accuracy":
+        return np.mean(valid_acc)
+    elif metric == "loss":
+        return np.mean(valid_loss)
+    else:
+        raise Exception("metric not defined as accuracy or loss")
```
Comment (d33bs) [inline 2.MLP_Multi_Classifier/MLP_utils/utils.py:546]: Consider providing a more specific Exception which could be used for this `raise`. Providing greater specificity here would help with readability, especially when/if this occurs.

File: 2.MLP_Multi_Classifier/MLP_utils/utils.py [hunk 7]
```diff
@@ -0,0 +1,938 @@
...
+        out_features = trial.suggest_int(
+            "n_units_l{}".format(i), params.LAYER_UNITS_MIN, params.LAYER_UNITS_MAX
+        )
+
+        layers.append(nn.Linear(in_features, out_features))
+        # activation function
+        layers.append(nn.ReLU())
+        # dropout rate
+        p = trial.suggest_float(
+            ("dropout_{}".format(i)), params.DROPOUT_MIN, params.DROPOUT_MAX
```
Comment (d33bs) [inline 2.MLP_Multi_Classifier/MLP_utils/utils.py:231]: Consider using an f-string here instead of `str.format()`.

File: 1.MLP_Binary_Classifier/MLP_utils/config.ini
Comment (d33bs) [inline 1.MLP_Binary_Classifier/MLP_utils/config.ini:1]: Seeing these `ini` files reminds me a bunch of `toml` formatting. Would using `toml` files offer any benefits here, keeping in mind that Python projects are beginning to use `pyproject.toml` files more commonly (recognizing that this content doesn't have to reside in the same place)?

**Review by d33bs — APPROVED** [review]
Nice work! I mostly had suggestions with my comments. LGTM! Please don't hesitate to let me know if you have any questions about anything.

File: 2.MLP_Multi_Classifier/MLP_utils/utils.py (hunk 1 above)
Comment (MikeLippincott) [reply 2.MLP_Multi_Classifier/MLP_utils/utils.py:70]: @d33bs Yes, there is and I have incorporated it in the `Parameters.py` filebut unforntunatly the parser retruns `str` type and I must type cast these into `int` or `float`.

File: 1.MLP_Binary_Classifier/nbconverted/Hyperparameter_Optimization_binary_class.py [hunk 8]
```diff
@@ -136,9 +136,9 @@
 X_train, X_test, X_val, Y_train, Y_test, Y_val = data_split(
     X_vals=df_values_X,
     y_vals=df_values_Y,
-    train=0.8,
-    val=0.1,
-    test=0.1,
+    train_proportion=0.8,
+    val_proportion=0.1,
+    test_proportion=0.1,
```
[the remark this answers is not in the document]
Comment (MikeLippincott) [reply 1.MLP_Binary_Classifier/nbconverted/Hyperparameter_Optimization_binary_class.py:141]: implemented

File: 2.MLP_Multi_Classifier/MLP_utils/utils.py (hunk 4 above)
Comment (MikeLippincott) [reply 2.MLP_Multi_Classifier/MLP_utils/utils.py:778]: addressed and changed

File: 2.MLP_Multi_Classifier/MLP_utils/utils.py (hunk 7 above)
Comment (MikeLippincott) [reply 2.MLP_Multi_Classifier/MLP_utils/utils.py:231]: fixed! :D

File: 2.MLP_Multi_Classifier/MLP_utils/config.ini
[the remark this answers is not in the document]
Comment (MikeLippincott) [reply 2.MLP_Multi_Classifier/MLP_utils/config.ini:1]: I created an issue for this as I will address when I merge the model files

File: 1.MLP_Binary_Classifier/MLP_utils/config.ini
Comment (MikeLippincott) [reply 1.MLP_Binary_Classifier/MLP_utils/config.ini:1]: toml added and ini dropped

File: 2.MLP_Multi_Classifier/nbconverted/Hyperparameter_Optimization_multiclass.py [hunk 9]
```diff
@@ -0,0 +1,291 @@
...
+# In[13]:
+
+
+plot_metric_vs_epoch(training_stats, "epochs_ran", "train_acc", "valid_acc")
+
+
+# In[14]:
+
+
+plot_metric_vs_epoch(training_stats, "epochs_ran", "train_loss", "valid_loss")
```
[the remark this answers is not in the document]
Comment (MikeLippincott) [reply 2.MLP_Multi_Classifier/nbconverted/Hyperparameter_Optimization_multiclass.py:269]: title and axis titles added!

File: 2.MLP_Multi_Classifier/MLP_utils/utils.py (hunk 1 above)
Comment (d33bs) [reply 2.MLP_Multi_Classifier/MLP_utils/utils.py]: Sorry to hear, I understand, thanks!
